### ISSUE TITLE
Real time reports with SSL

### ISSUE THREAD

**Issue by fabioneves** [allinurl/goaccess#474]
Hello,

I've tried to run a simple real time report like this:
goaccess -f access.log -o report.html --real-time-html --ws-url=goaccess.io

The problem is that my nginx vhost is running in SSL and I get the following error on the console:
report.html:1111 Mixed Content: The page at 'https://testing.com/report.html' was loaded over HTTPS, but attempted to connect to the insecure WebSocket endpoint 'ws://testing.com:7890/'. This request has been blocked; this endpoint must be available over WSS.setWebSocket @ report.html:1111
report.html:1111 Uncaught SecurityError: Failed to construct 'WebSocket': An insecure WebSocket connection may not be initiated from a page loaded over HTTPS.

I searched in the documentation for SSL and HTTPS but couldn't find anything, so for now I'm limited to HTTP?

Thanks!

**Comment by allinurl** [comment]
Hello,

That's correct. SSL is not yet supported. I need to implement this to the [websocket server](https://github.com/allinurl/gwsocket). Implementing this shouldn't be too bad, it would probably require openssl.

Stay tuned!

**Comment by odigity** [comment]
Just learned about and tried out GoAccess for the first time tonight; ran into the same problem.  Definitely looking forward to this feature being added, as I really like what I've seen so far, but don't want to have to support non-SSL access to the report.

**Comment by coderonline** [comment]
+1 for wss://

- and thanks for goaccess! For me it has become the greatest solution for that kind with the 1.0 release and I love it's look, performance and the fact, that it is written in C :)

**Comment by mitchellkrogza** [comment]
Thanks for GoAccess, such a well designed and useful utility for watching my Nginx logs.

Please could you add a note on the documentation at https://goaccess.io/man#examples and https://goaccess.io/faq that only http can be used for realtime reporting. 

This had me going in circles for a while until I found this issue and realised I had to just setup a separate domain that runs on http with all .html and realtime reports located there. 

Static reports work fine over SSL but not the realtime.

**Comment by stefanmb** [comment]
Just wanted to chime in as well - GoAccess is a fantastic piece of software! Thank you!

I experimented with reverse-proxying the realtime endpoint over Nginx with TLS but was unable to get it working, my understanding is `wss:/` is needed so some modifications would have to be made to the generated HTML for it to work, at the very least.

**Comment by allinurl** [comment]
@mitchellkrogza Thanks. I have added the notes to those two pages.

@stefanmb I've been working on this so I expect to have `wss://` support soon. I may need to push out a minor release before deploying this change though.

**Comment by evemilano** [comment]
Hello, I'm using goaccess to parse a log from an https server, and to show real time report over an other not-https website. The report is loaded with data, but it seems like a static one, infact it has a flag "Disconnected". Is this related to HTTPS issue? Or am I missing something?

**Comment by allinurl** [comment]
@evemilano If your report is being served through an https‎ address such as https://site.com/report.html you won't be able to use the real time functionality at the moment. If it's through an http site, then it should work fine.

**Comment by evemilano** [comment]
the report is published in http://app.example.com/report.html and it takes the log from https://example.com. Should it work? Because it doesn't :(

**Comment by allinurl** [comment]
If it's showing as disconnected, then sounds like your browser is not able to establish a connection with the WebSocket server. Please take a look at your browser’s console and post back any errors there may be.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        From: Giovanni SacheliSent: Saturday, August 20, 2016 10:09 PMTo: allinurl/goaccessReply To: allinurl/goaccessCc: Gerardo O.; CommentSubject: Re: [allinurl/goaccess] Real time reports with SSL (#474)the report is published in http://app.example.com/report.html and it takes the log from https://example.com. Should it work? Because it doesn't :(

—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/allinurl/goaccess","title":"allinurl/goaccess","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/allinurl/goaccess"}},"updates":{"snippets":[{"icon":"PERSON","message":"@evemilano in #474: the report is published in http://app.example.com/report.html and it takes the log from https://example.com. Should it work? Because it doesn't :("}],"action":{"name":"View Issue","url":"https://github.com/allinurl/goaccess/issues/474#issuecomment-241221046"}}}

**Comment by evemilano** [comment]
the console says: 
`WebSocket connection to 'ws://www.mysite.com:7890/' failed: Error in connection establishment: net::ERR_CONNECTION_TIMED_OUT`

I continue here the discussion :)

**Comment by madpsy** [comment]
A note for a quick workaround until wss is implemented.

In Chrome you can just click the shield and 'allow unsafe scripts' but in Firefox you need to go to 'about:config' and then set 'network.websocket.allowInsecureFromHTTPS' to true.

**Comment by aslian** [comment]
If anyone interested... 
Have a workaround for that.

* Let's say you have HTTPS site on **nginx** with HSTS enabled.
* Let's say goaccess write its live report file to /report.html (in your site root folder).

Then,

```
server {
        listen 80;
        server_name my.site.com;
        
        location /report.html {
                add_header Strict-Transport-Security "max-age=0;";
                root /var/www/mysite;    
        }
        
        location / {
                return 302 https://$host$request_uri;
        }
}

server {
    listen 443;
    server_name my.site.com;

    root /var/www/mysite;
    index index.php index.html;

    ssl on;
	...
    location /report.html {
        add_header Strict-Transport-Security "max-age=0;";
        return 302 http://$host$request_uri;  # <-- Here's a trick
    }
	...
}
```
It will downgrade connection to HTTP for that single file.

**Comment by allinurl** [comment]
TLS/SSL has been implemented. Feel free to build from [**development**](https://github.com/allinurl/goaccess#build-from-github-development) for testing this out.

It will be pushed out in the upcoming version (which I expect to release soon).

**Comment by allinurl** [comment]
Just for the record, it needs to be built as:

```
$ ./configure --enable-geoip --enable-utf8 --with-openssl
$ make
# make install
```

and run as:

    # goaccess -f access.log -o report.html --real-time-html --ws-url=<host> --ssl-cert=<cert> --ssl-key=<priv.key>

**Comment by madpsy** [comment]
Superb stuff - shall give this a go when I have a bit of time.

**Comment by allinurl** [comment]
This has been shipped in [**v1.1**](https://goaccess.io/) which is out now :)
